Elevation measure widget

jailln opened the pull request:
## Description
Add a widget to measure elevation anywhere in the 3D scene. Also contains a few modifications of the API (`Picking` and `View`).

I will write tests and squash commits once we agree on the implementation :) (i.e. after a first review)

Note that I opened a [PR on itowns2-sample-data](https://github.com/iTowns/iTowns2-sample-data/pull/8) for the assets related to this widget (sprite, logo and dataset for the example). 

## Motivation and Context
Implementation of the first tool described in #1868

## Screenshots (if appropriate)

Mesure elevation on the terrain:

![image](https://user-images.githubusercontent.com/16967916/183654952-918a9b0c-2cef-4a22-9700-5a36a8172521.png)

Mesure elevation on any 3D object:
![image](https://user-images.githubusercontent.com/16967916/183655203-47b88ad7-dee5-4e44-810b-905018dd6ebd.png)

jailln committed:
fix(picking): fix picking on multiple layers

mgermerie commented:
There is an issue with the Elevation widget : if you use it while zoomed out, the elevation is not red on the most detailed tile (since it is not necessarily loaded yet). The value will therefore differ on the same point, whether you use the widget while zoomed out or zoomed in at maximum.

mgermerie commented on src/Core/View.js:
```diff
@@ -1057,9 +1057,9 @@ class View extends THREE.EventDispatcher {
+     * @returns {Coordinates}   The world {@link Coordinates} of the terrain at the given view coordinates.
      */
-    pickCoordinates(mouse, target = new Coordinates(this.tileLayer.extent.crs)) {
+    pickTerrainCoordinates(mouse, target = new Coordinates(this.referenceCrs)) {
```
Renaming this method introduces a breaking change. Could you still define `pickCoordinates` with a deprecation warning and a calling to the new method ?

jailln replied:
Yep and I will also add a breaking change in the commit message.

mgermerie commented on src/Core/View.js:
```diff
@@ -1072,9 +1072,7 @@ class View extends THREE.EventDispatcher {
-        coordinates.crs = this.referenceCrs;
-        coordinates.setFromVector3(positionVector);
-        coordinates.as(target.crs, target);
+        target.setFromVector3(positionVector);
```
We need to leave the transformation here, since use should be able to call the method with any `target` coordinate. We can't assume the CRS of `target`. For example if I type : 
```js
const coordinates = new itowns.Coordinates('EPSG:4326');
view.pickTerrainCoordinates({ /* some mouse position */ }, coordinates);
```
it will return some coordinates with an `EPSG:4326` CRS but whose `x`, `y` and `z` values are intended as `EPSG:4978`, so most probably not the coordinates under the mouse position.

jailln replied:
yep you're right, I got mixed up :)

mgermerie commented on src/Core/View.js:
```diff
@@ -1057,9 +1057,9 @@ class View extends THREE.EventDispatcher {
+     * @returns {Coordinates}   The world {@link Coordinates} of the terrain at the given view coordinates.
      */
-    pickCoordinates(mouse, target = new Coordinates(this.tileLayer.extent.crs)) {
+    pickTerrainCoordinates(mouse, target = new Coordinates(this.referenceCrs)) {
```
We can indeed change the default target CRS with the view's reference CRS, as it seems more logical.
However we should keep these two following points in mind : 
- this introduces a breaking change for some users (who for example feed the default output of this method to other processes without transforming it).
- In a `PlanarView` context, I believe that `view.referenceCrs` and `view.tileLayer.extent.crs` are always the same. Having the default output CRS to `view.tileLayer.extent.crs` allows to output coordinates in longitude, latitude and altitude, whose representation is more easy to understand than coordinates in `EPSG:4978`.

Therefore, do you still think we should change it ? In any case, could you please add the previously missing documentation on the default output CRS ?

jailln replied:
I think yes because it is clearer for users since in my opinion users don't really know the crs of the extent of the tileLayer which is quite internal to itowns, they should expect to get coordinates in the view CRS which they actually know of. You're right that it is a breaking change, I will add one.

mgermerie commented on src/Core/Geographic/Coordinates.js:
```diff
@@ -58,15 +58,19 @@ class Coordinates {
      * @param {number} [v2=0] - z or altitude value.
      */
-    constructor(crs, v0 = 0, v1 = 0, v2 = 0) {
+    constructor(crs = 'EPSG:4978', v0 = 0, v1 = 0, v2 = 0) {
```
The documentation is a good addition, thanks !
On another hand, I don't think we should assume a default CRS for coordinates. No coordinates make sense without a CRS and in my opinion, user should always have to specify the CRS associated with the coordinates values he is dealing with.

jailln replied:
Yes I agree with you, I will remove the default CRS.

mgermerie commented on src/Core/Geographic/Coordinates.js:
```diff
@@ -92,6 +96,15 @@ class Coordinates {
+     * Sets the Coordinate Reference System.
+     * @param {String} crs Coordinate Reference System (e.g. 'EPSG:4978')
+     */
+    setCrs(crs) {
```
Good idea to add this check of parameter validity !

mgermerie commented on src/Core/Picking.js:
```diff
@@ -187,9 +188,16 @@ export default {
                 for (let i = 0; i < candidates.length; i++) {
                     if (candidates[i].objId == o.baseId) {
+                        // Compute distance to the camera
+                        const pointCoordinate = new Coordinates(view.referenceCrs, o.position.x, o.position.y, o.position.z);
```
I think this method for computing distance could cause issues : the `position` property for threejs `Object3D` is expressed in the object's parent coordinates system.
Here, the point could be centered within a `Group`, and the `Group` could be positioned at the correct world coordinates. the `position` of the point would therefore be `(0, 0, 0)`. Using world position for the point, as you did for the camera would prevent this.

mgermerie replied:
Also, do you know of an usage of this picking method other than with `PointCloudLayer` ?

jailln replied:
Yes, you're right, I added the world transformation and updated a bit this part to fix a bug that I had not seen for point cloud picking.

jailln committed:
example(potree_25d_map): declare map projection code

jailln committed:
refacto(view): rename pickCoordinates to pickTerrainCoordinates

BREAKING CHANGE: View.pickCoordinates has been renamed to View.pickTerrainCoordinates and returns the coordinates in the referenceCrs of the view instead of in the crs of the tiledLayer extent.

jailln committed:
feat(picking): Add distance and point position to returned object by points picking

jailln committed:
feat(Coordinates): Add a setCrs method

jailln committed:
feat(widgets): Add a widget to measure elevation at any point in the 3D scene

jailln commented:
@mgermerie I made the modifications you recommended, squashed and rebased.

jailln commented:
Some functional tests don't pass (timeout), while they pass locally. I tried to update a few things but it didn't change anything. Any ideas @mgermerie @gchoqueux ?

jailln commented:
I won't have time to finish this feature in the short term, closing for now.
